### ISSUE TITLE
Rule proposal: `no-void-return-type`

### ISSUE THREAD

**Issue by remcohaszing** [typescript-eslint/typescript-eslint#7250]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+new+plugin+rule%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#supported-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] My proposal specifically checks TypeScript syntax, or it proposes a check that requires type information to be accurate.
- [X] My proposal is not a "formatting rule"; meaning it does not just enforce how code is formatted (whitespace, brace placement, etc).
- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Description

The `void` type in TypeScript shouldn’t exist, because it represents something that doesn’t exist in JavaScript. https://github.com/microsoft/TypeScript/issues/42709 explains it in more depth.

There is are the [`@typescript-eslint/no-invalid-void-type`](https://typescript-eslint.io/rules/no-invalid-void-type) and [`@typescript-eslint/no-meaningless-void-operator`](https://typescript-eslint.io/rules/no-meaningless-void-operator) rules. I would like to propose the `@typescript-eslint/no-void-return-type` rule to disallow returning `void`, `Promise<void>`, and `PromiseLike<void>`.

### Fail Cases

```typescript
// This function returns undefined
function a(): void {}

// This function returns undefined
function b() {}

// This probably types a function that returns undefined
declare function c(): void

// This function returns a promise that resolves to undefined
async function d(): Promise<void> {}

// This function returns a promise that resolves to undefined
declare function e(): Promise<void>

// This function returns a promise-like that resolves to undefined
declare function e(): PromiseLike<void>

// If void is used as anything but a return value, it’s invalid anyway
type G = string | void
```

### Pass Cases

```typescript
// This function returns undefined
function a(): undefined {}

// This function returns undefined
function b(): undefined {}

// This probably types a function that returns undefined
declare function c(): undefined

// This function returns a promise that resolves to undefined
async function d(): Promise<undefined> {}

// This function returns a promise that resolves to undefined
declare function e(): Promise<undefined>

// This function returns a promise-like that resolves to undefined
declare function e(): PromiseLike<undefined>
```

### Additional Info

[TypeScript 5.1](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-5-1.html) loosened the rules for `void`/`undefined`, making it viable to stop using `void`.

This rule could be auto-fixable by replacing a return type of `void` with `undefined`, or adding an explicit return type if none of the branches return an explicit value.

It would also just report any usage of `void`.

**Comment by remcohaszing** [comment]
Thanks for the tip to use `no-restricted-syntax`. It doesn’t catch implicit `void` return types, but it’s sufficient for my use case.
